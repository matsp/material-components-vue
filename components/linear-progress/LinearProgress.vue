<template>
  <div
    role="progressbar"
    class="mdc-linear-progress"
  >
    <div class="mdc-linear-progress__buffering-dots" />
    <div class="mdc-linear-progress__buffer" />
    <div class="mdc-linear-progress__bar mdc-linear-progress__primary-bar">
      <span class="mdc-linear-progress__bar-inner" />
    </div>
    <div class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar">
      <span class="mdc-linear-progress__bar-inner" />
    </div>
  </div>
</template>

<script>
import { MDCLinearProgress } from '@material/linear-progress'

import { baseComponentMixin, themeClassMixin } from '@components/base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    open: {
      type: Boolean,
      default: false
    },
    indeterminate: {
      type: Boolean,
      default: false
    },
    reverse: {
      type: Boolean,
      default: false
    },
    progress: {
      type: Number,
      validator: (value) => (value >= 0) && (value <= 1),
      default: 1
    },
    buffer: {
      type: Number,
      validator: (value) => (value >= 0) && (value <= 1),
      default: 1
    }
  },
  data () {
    return {
      mdcLinearProgress: undefined
    }
  },
  watch: {
    progress () {
      this.mdcLinearProgress.progress = this.progress
    },
    buffer () {
      this.mdcLinearProgress.buffer = this.buffer
    },
    indeterminate () {
      this.mdcLinearProgress.determinate = !this.indeterminate
    },
    reverse () {
      this.mdcLinearProgress.reverse = this.reverse
    },
    open () {
      this.open ? this.mdcLinearProgress.open() : this.mdcLinearProgress.close()
    }
  },
  mounted () {
    this.mdcLinearProgress = MDCLinearProgress.attachTo(this.$el)
    this.mdcLinearProgress.determinate = !this.indeterminate
    this.mdcLinearProgress.reverse = this.reverse
    this.mdcLinearProgress.progress = this.progress
    this.mdcLinearProgress.buffer = this.buffer
    this.open ? this.mdcLinearProgress.open() : this.mdcLinearProgress.close()
  },
  beforeDestroy () {
    this.mdcLinearProgress.destroy()
  }
}
</script>
