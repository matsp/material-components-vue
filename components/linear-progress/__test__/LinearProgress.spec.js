import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import LinearProgress from '../LinearProgress.vue'

describe('LinearProgress', () => {
  it('should mount', () => {
    const wrapper = mount(LinearProgress)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.mdcLinearProgress).toBeDefined()
  })

  it('should render with no prop', () => {
    const wrapper = mount(LinearProgress)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-linear-progress')
    expect(wrapper.classes()).toContain('mdc-linear-progress--closed')
  })

  it('should render as open', () => {
    const wrapper = mount(LinearProgress, {
      propsData: {
        open: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.isVisible()).toBe(true)
  })

  it('should render as reversed', () => {
    const wrapper = mount(LinearProgress, {
      propsData: {
        open: true,
        reverse: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-linear-progress--reversed')
  })

  it('should render as indeterminate', () => {
    const wrapper = mount(LinearProgress, {
      propsData: {
        open: true,
        indeterminate: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-linear-progress--indeterminate')
  })
})
