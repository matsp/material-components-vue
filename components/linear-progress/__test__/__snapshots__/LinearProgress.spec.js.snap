// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinearProgress should render as indeterminate 1`] = `
<div role="progressbar" aria-label="" aria-valuemin="0" aria-valuemax="1" class="mdc-linear-progress mdc-linear-progress--indeterminate">
  <div class="mdc-linear-progress__buffering-dots"></div>
  <div class="mdc-linear-progress__buffer" style="transform: scaleX(1);"></div>
  <div class="mdc-linear-progress__bar mdc-linear-progress__primary-bar" style="transform: scaleX(1);"><span class="mdc-linear-progress__bar-inner"></span></div>
  <div class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"><span class="mdc-linear-progress__bar-inner"></span></div>
</div>
`;

exports[`LinearProgress should render as open 1`] = `
<div role="progressbar" aria-label="" aria-valuemin="0" aria-valuemax="1" aria-valuenow="1" class="mdc-linear-progress">
  <div class="mdc-linear-progress__buffering-dots"></div>
  <div class="mdc-linear-progress__buffer" style="transform: scaleX(1);"></div>
  <div class="mdc-linear-progress__bar mdc-linear-progress__primary-bar" style="transform: scaleX(1);"><span class="mdc-linear-progress__bar-inner"></span></div>
  <div class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"><span class="mdc-linear-progress__bar-inner"></span></div>
</div>
`;

exports[`LinearProgress should render as reversed 1`] = `
<div role="progressbar" aria-label="" aria-valuemin="0" aria-valuemax="1" aria-valuenow="1" class="mdc-linear-progress mdc-linear-progress--reversed">
  <div class="mdc-linear-progress__buffering-dots"></div>
  <div class="mdc-linear-progress__buffer" style="transform: scaleX(1);"></div>
  <div class="mdc-linear-progress__bar mdc-linear-progress__primary-bar" style="transform: scaleX(1);"><span class="mdc-linear-progress__bar-inner"></span></div>
  <div class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"><span class="mdc-linear-progress__bar-inner"></span></div>
</div>
`;

exports[`LinearProgress should render with no prop 1`] = `
<div role="progressbar" aria-label="" aria-valuemin="0" aria-valuemax="1" aria-valuenow="1" class="mdc-linear-progress mdc-linear-progress--closed">
  <div class="mdc-linear-progress__buffering-dots"></div>
  <div class="mdc-linear-progress__buffer" style="transform: scaleX(1);"></div>
  <div class="mdc-linear-progress__bar mdc-linear-progress__primary-bar" style="transform: scaleX(1);"><span class="mdc-linear-progress__bar-inner"></span></div>
  <div class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"><span class="mdc-linear-progress__bar-inner"></span></div>
</div>
`;
